### PR TITLE
Add Gloo Gather Function

### PR DESCRIPTION
**Pull request by isky-cd**
### PR types

New features
### PR changes

OPs
### Describe

Add Gloo Gather Function

### CODE REVIEW

**Review by kuizhiqing — DISMISSED** [review]
lgtm

File: python/paddle/fluid/tests/unittests/collective/process_group_gloo.py [hunk 1]
```diff
@@ -204,6 +204,30 @@ def test_create_process_group_gloo(self):
...
+            if pg.rank() == root:
+                task = pg.gather(tensor_x, tensor_y[root], root, True)
+                task.wait()
+                assert np.array_equal(tensor_x, gather_result)
+            else:
+                task = pg.gather(tensor_x, tensor_y[1 - root], root, True)
+                task.wait()
+
+        test_gather(0)
+        test_gather(1)
```
Comment (sneaxiy) [inline python/paddle/fluid/tests/unittests/collective/process_group_gloo.py]: You assume that this ut would only run when world size is 2. If world size is not 2, this ut does not work.

File: paddle/fluid/distributed/collective/gloo_tools.h [hunk 2]
```diff
@@ -61,8 +61,6 @@ class SendRecvOptions {
 
   friend void send_recv(SendRecvOptions*);
 };
-
```
Comment (sneaxiy) [inline paddle/fluid/distributed/collective/gloo_tools.h]: You do not need to rename this file. Rename it back.

Comment (isky-cd) [reply paddle/fluid/distributed/collective/gloo_tools.h]: Done

File: python/paddle/fluid/tests/unittests/collective/process_group_gloo.py (hunk 1 above)
Comment (isky-cd) [reply python/paddle/fluid/tests/unittests/collective/process_group_gloo.py]: Done

File: paddle/fluid/distributed/collective/process_group_gloo.cc [hunk 3]
```diff
@@ -680,6 +681,61 @@ std::shared_ptr<ProcessGroup::Task> ProcessGroupGloo::Scatter(
...
+    gloo::gather(opts);
+  }
+};
+
+std::shared_ptr<ProcessGroup::Task> ProcessGroupGloo::Gather(
+    phi::DenseTensor* out_tensor,
+    const phi::DenseTensor& in_tensor,
+    const GatherOptions& opts,
+    bool sync_op,
+    bool use_calc_stream) {
```
Comment (sneaxiy) [inline paddle/fluid/distributed/collective/process_group_gloo.cc:729]: Throw exception when `use_calc_stream` is true.

File: paddle/fluid/distributed/collective/process_group_gloo.h [hunk 4]
```diff
@@ -210,6 +216,11 @@ class ProcessGroupGloo : public ProcessGroupWithoutStream {
     return platform::DeviceContextPool::Instance().Get(place);
   }
 
+  phi::DeviceContext* GetDeviceContext(const Place& place,
```
Comment (sneaxiy) [inline paddle/fluid/distributed/collective/process_group_gloo.h:219]: Do not need this method.

File: paddle/fluid/distributed/collective/process_group_gloo.cc (hunk 3 above)
Comment (isky-cd) [reply paddle/fluid/distributed/collective/process_group_gloo.cc:729]: Done

File: paddle/fluid/distributed/collective/process_group_gloo.h (hunk 4 above)
Comment (isky-cd) [reply paddle/fluid/distributed/collective/process_group_gloo.h:219]: An exception will be thrown if this method is not used.

Comment (isky-cd) [reply paddle/fluid/distributed/collective/process_group_gloo.h:219]: I add an exception when use_calc_stream is true.

**Review by sneaxiy — APPROVED** [review]

**Pull request merged**